### DISCUSSION TITLE
Question to Kafka Metrics

### DISCUSSION THREAD

**Discussion by bartman64** [confluentinc/parallel-consumer#368] [Q&A] [answered]
So we are using grafana to get a glance of our services and kafka metrics. Those services were implement on quarkus base using kafka-streams.
Now with the switch to the parallel consumers the metrics are no longer visible in grafana. Currently I don't know, if with some config properties I could make them visible again or those aren't implemented/usable yet since I saw a open issues regarding metrics (https://github.com/confluentinc/parallel-consumer/issues/27)
I understand that kafka-streaming differs to the kafka low-level-api approach used from the parallel consumer, but I don't know how to access those metrics.

If someone could give me some insight would be great, thx 😄

**Comment by astubbs** [comment, marked as answer]
#27 is just about adding extra metrics specific to parallel consumer. 

You should still be able to access all the metric exposed by the consumer and producer instances - can you not see those?

**Reply by bartman64** [reply]
Hmm ok...
So we have a grafana dashboard with all those metrics and before switching to the parallel consumer, we were able to see the metrics.
After that the graphs stay empty.
So I was just wondering what happend there, but havn't looked deeper into that myself

**Comment by amolkamble17** [comment]
We are having quarkus streaming application with pre distribution pipeline which produces messages to multiple topics
and parallel consumer (non streaming application) consumes that message and writes to output topic. We are not able to see metrics in Grafana dashboard of non streaming application when it consumes messages produced by Kafka streaming.

![metrics](https://user-images.githubusercontent.com/94970768/195058534-8c44518b-7bbe-4d8c-94b4-8a4c1c51317b.png)

**Reply by astubbs** [reply]
Sorry @amolkamble17, I missed your question. I'm not sure I follow - are you saying you can't see any JMX sourced metrics from PC's consumer or producer instances?

FYI, keep an eye on: 
- #485

**Reply by amolkamble17** [reply]
Thank you @astubbs for reply.
Yes, we are using parallel consumer as below.

var options = ParallelConsumerOptions.<String, String>builder()
                .ordering(KEY)
                .commitMode(ParallelConsumerOptions.CommitMode.PERIODIC_CONSUMER_ASYNCHRONOUS)
                .maxConcurrency(Integer.parseInt(10))
                .consumer(kafkaConsumer)
                .producer(kafkaProducer)
                .build();

        ParallelStreamProcessor<String, String> eosStreamProcessor = ParallelStreamProcessor.createEosStreamProcessor(options);
        eosStreamProcessor.subscribe(of(inputTopic + "." + randomNumber));

1. How do we get metrics of parallel consumer to show on Grafana dashboard.

We are having quarkus stream application which already expose kafka metrics for streaming application but it is not showing metrics of parallel consumer.  Currently we are using 0.5.2.4 version of parallel-consumer-core. 
Could you please give some example how to extract metrics for parallel consumer and that would be available with existing metrics.  

2. Which version and when this enhancement of PC metrics will be complete.

**Comment by astubbs** [comment]
@amolkamble17, @bartman64 - the metrics were never gone; they were unbound, and the maintained fork's README now shows the wiring.

Two sets of meters are involved. The Kafka client's own metrics are JMX on the `KafkaConsumer` and `KafkaProducer` you hand to PC, and a framework's Kafka binder - Quarkus's, Spring's - binds the clients *it* creates, not one you built yourself, which is the most likely reason the graphs went empty when you switched. Bind yours explicitly with Micrometer's `KafkaClientMetrics` on the consumer instance you pass in. The second set is Parallel Consumer's own Micrometer meters - in-flight work, shards, commits, the load gate - which are recorded into the `MeterRegistry` you set on `ParallelConsumerOptions` and are a no-op if you set none. The README's "Metrics" section lists the meters and "Example Metrics setup steps" shows both bindings on one registry, with the `pcinstance` tag for the case of several instances in one JVM. Both are on [v0.6.0.0](https://github.com/astubbs/parallel-consumer/releases/tag/v0.6.0.0).


The coordinates changed with the fork - on Maven Central it is now:

```xml
<dependency>
    <groupId>bz.stub.parallelconsumer</groupId>
    <artifactId>parallel-consumer-core</artifactId>
    <version>0.6.0.0</version>
</dependency>
```

(`parallel-consumer-vertx`, `parallel-consumer-reactor` and `parallel-consumer-mutiny` likewise.) The Java package is `bz.stub.parallelconsumer` too, so imports change with it - one command rewrites them (GNU `sed`; on macOS use `sed -i ''`):

```bash
find . -name '*.java' -exec sed -i 's/io\.confluent\.parallelconsumer/bz.stub.parallelconsumer/g' {} +
```
